R/comparisons.R

```diff
@@ -0,0 +1,239 @@
+#' pandemic_flu <- scenario(
+#'   model_function = "finalsize::final_size",
+#'   parameters = make_parameters_finalsize_UK(r0 = 1.5),
+#'   replicates = 1L
+#' )
+#'
+#' covid19 <- scenario(
+#'   model_function = "finalsize::final_size",
+#'   parameters = make_parameters_finalsize_UK(r0 = 5.0),
+#'   replicates = 1L
+#' )
+#'
+#' # create a comparison object
+#' comparison(
+#'   pandemic_flu = pandemic_flu, covid19 = covid19,
+#'   baseline = "pandemic_flu"
+#' )
+#'
+#' # pass scenario objects as a list
+#' # create a comparison object
+#' comparison(
+#'   list(pandemic_flu = pandemic_flu, covid19 = covid19),
+#'   baseline = "pandemic_flu"
+#' )
+comparison <- function(...,
+                       baseline,
+                       match_variables,
+                       comparison_variables) {
+  # check input
+  data <- list(...)
+  if ((length(data) == 1L) && (is.list(data[[1]])) &&
+    (!is.scenario(data[[1]]))) {
+    data <- data[[1]]
+  }
+  if (missing(match_variables)) {
+    match_variables <- NA_character_
+  }
+  if (missing(comparison_variables)) {
+    comparison_variables <- NA_character_
+  }
+
+  stopifnot(
+    "All objects must be of the 'scenario' class" =
+      all(
+        vapply(
+          data, is.scenario,
+          FUN.VALUE = TRUE
+        )
+      ),
+    "Baseline must be among scenario names" =
+      (is.character(baseline) && baseline %in% names(data)),
+    "Matching variables must be a string" =
+      (is.character(match_variables)),
+    "Comparison variables must be a string" =
+      (is.character(comparison_variables))
+  )
+
+  # call comparison constructor
+  object <- new_comparison(
+    data, baseline, match_variables, comparison_variables
+  )
+
+  # call comparison validator
+  validate_comparison(object)
+
+  # return comparison object
+  object
+}
+
+#' Validator for the 'comparison' class
+#'
+#' @param object A 'comparison' object.
+#'
+#' @return None. Errors when an invalid 'comparison' object is provided.
+validate_comparison <- function(object) {
+  # check for class and class invariants
+  stopifnot(
+    "Object should be of class comparison" =
+      (is.comparison(object)),
+    "'comparison' object does not contain the correct attributes" =
+      (all(
+        c(
+          "data", "baseline"
+        ) %in% attributes(object)$names
+      )
+      ),
+    "Comparison must be a list of 'scenario' objects" =
+      (is.list(object$data) && (
+        all(
+          vapply(
+            object$data, is.scenario,
+            FUN.VALUE = TRUE
+          )
+        )
+      )),
+    "Baseline must be among scenario names" =
+      (is.character(object$baseline) &&
+        object$baseline %in% names(object$data))
+  )
+  invisible(object)
+}
+
+#' @export
+print.comparison <- function(x, ...) {
+  # prepare information
+  header <- cli::style_bold("Scenario comparison object")
+  scenario_count <- glue::glue(" # Scenarios: {length(x$data)}")
```

Comment:
`#` is a bit ambiguous here, would recommend changing.

Reply:
Fixed in https://github.com/epiverse-trace/scenarios/pull/12/commits/38bc11dd958f58071584870b82c2fba302aa7c19